### PR TITLE
Wallet support for bootstrapping

### PR DESCRIPTION
**Pull request by aguycalled**
This PR adds support for bootstrapping the blockchain directly from the wallet.

A new argument (-bootstrap) is handled on initialisation to specify an URL from where a copy of the blockchain in TAR format will be downloaded and extracted in the data folder.

Adds too in the GUI a submenu entry under FILE to do this operation graphically.

### CODE REVIEW

**Comment by red010b37** [comment]
Sorry - again don't want to release this in 4.2.0, but could be its own patch release

File: src/init.cpp [hunk 1]
```diff
@@ -800,6 +831,63 @@ void InitLogging()
  */
 bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 {
+    // ********************************************************* Step 0: download bootstrap
```
Comment (red010b37) [inline src/init.cpp]: Does this mean it runs on every start? as its part of the App Init

File: src/init.cpp [hunk 2]
```diff
@@ -800,6 +831,63 @@ void InitLogging()
...
+                        boost::filesystem::remove_all(sDownload);
+                        if(!fOk)
+                            return InitError("[BOOTSTRAP] Could not extract data.");
+                    }
+                } else {
+                    return InitError("[BOOTSTRAP] Failed!");
+                }
+            }
+        } else {
+            return InitError("[BOOTSTRAP] Failed!");
```
Comment (red010b37) [inline src/init.cpp]: Should this be more descriptive, as it's not that the bootstrap failed its the lib doesn't exist? so we exit. Thin it should say, Failed as curl is not found

File: src/init.cpp [hunk 3]
```diff
@@ -800,6 +831,63 @@ void InitLogging()
...
+    if(sBootstrap != "")
+    {
+        RemoveConfigFile("bootstrap");
+        uiInterface.InitMessage(strprintf("[BOOTSTRAP] Trying to download from %s", sBootstrap.c_str()));
+        CURL *curl;
+        FILE *fp, *a;
+        CURLcode res;
+
+        curl = curl_easy_init();
+        if (curl) {
```
Comment (red010b37) [inline src/init.cpp]: is curl os specific. e.g you need it installed on your machine, or is packaged in the app? - just asking

File: src/init.cpp (hunk 1 above)
Comment (aguycalled) [reply src/init.cpp]: yes

File: src/init.cpp (hunk 3 above)
Comment (aguycalled) [reply src/init.cpp]: the library is linked, curl is already used in other parts of navcoin-core

File: src/init.cpp (hunk 2 above)
Comment (aguycalled) [reply src/init.cpp]: if the lib does not exist, the code won't compile.
curl_easy_init( ) is NULL when "something went wrong" as per Curl official documentation (https://curl.haxx.se/libcurl/c/curl_easy_init.html) - hence the lack of definition in the warning text

File: src/init.cpp (hunk 3 above)
Comment (red010b37) [reply src/init.cpp]: thanks :)

File: src/init.cpp [hunk 4]
```diff
@@ -799,6 +830,63 @@ void InitLogging()
  */
 bool AppInit2(boost::thread_group& threadGroup, CScheduler& scheduler)
 {
+    // ********************************************************* Step 0: download bootstrap
+    std::string sBootstrap = GetArg("-bootstrap","");
+
+    if(sBootstrap != "")
```
Comment (red010b37) [inline src/init.cpp]: I think this should be split out to its own function

Comment (red010b37) [reply src/init.cpp]: Thoughts?

**Pull request merged**